tracer/src/tracer.rs

```diff
@@ -107,17 +107,15 @@ struct TraceEvent {
 
 pub fn trace_point_log(event: &'static str) {
     let trace_event = TraceEvent {
-        // SAFETY: start has been initialised as part of initialising the value of TRACER
-        timestamp: Instant::now().duration_since(unsafe { TRACER.get().unwrap().start }),
+        timestamp: Instant::now().duration_since(TRACER.get().unwrap().start),
         event,
         end_timestamp: None,
-        // SAFETY: thread_depth accesses current thread only specific data
-        depth: unsafe { TRACER.get().unwrap().thread_depth() },
+        depth: TRACER.get().unwrap().thread_depth(),
     };
-    // SAFETY: add_event accesses current thread only specific data
-    unsafe {
-        TRACER.get_mut().unwrap().add_event(trace_event);
-    }
+
+    let mut copy = TRACER.get().unwrap().clone();
+    copy.add_event(trace_event);
+    TRACER.set(copy).unwrap();
```

Comment:
This has a data race which could result in data being lost. 

e.g.
A copies
A updates
B copies
B updates
B sets
A sets

Changes made by B are lost.

Also deep cloning the hash tree is very costly.

The current solution avoids the race by having every thread update only it's own entry - however I guess there are rare races inside the `HashMap` implementation.

Reply:
I was thinking of using `arc_swap` for this (it has an `rcu` method), or stick with `once_cell` we use a `RwLock` to prevent data race (acquire a write lock then update).

I'll use `#[allow(static_mut_refs)]` for now then :)

Reply:
The goal of the tracing infrastructure was not to affect performance / thread ordering by having a lock. Also `Mutex` outperforms (historically at least) `RwLock` (https://www.slideshare.net/slideshow/performance-comparison-of-mutex-rwlock-and-atomic-types-in-rust/72692582#16)

Reply:
Thanks for elaborating, I have learned a lot! :)